src/printer.js

```diff
@@ -3260,6 +3272,89 @@ function hasLeadingOwnLineComment(text, node) {
   return res;
 }
 
+function hasNakedLeftSide(node) {
+  return (
+    node.type === "BinaryExpression" ||
+    node.type === "LogicalExpression" ||
+    node.type === "ConditionalExpression" ||
+    node.type === "CallExpression" ||
+    node.type === "MemberExpression" ||
+    node.type === "SequenceExpression" ||
+    node.type === "TaggedTemplateExpression"
+  );
+}
+
+function getLeftSide(node) {
+  if (node.expressions) {
+    return node.expressions[0];
+  }
+  return node.left || node.test || node.callee || node.object || node.tag;
+}
+
+function exprNeedsASIProtection(node) {
+  const maybeASIProblem = node.needsParens ||
```

Comment:
This is where `needsParens` is used. 

`exprNeedsASIProtection` is recursive, and recurses on the basis of the ast nodes via `getLeftSide`, just above. So it'll be a bit of a pain to move to using `path.needsParens()`, in addition to being slower. 

But this is pretty ugly. 

Thoughts?